File: 2331-evaluate-boolean-binary-tree/2331-evaluate-boolean-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool evaluateTree(TreeNode* root) {
        // Handle the case for leaf nodes.
        if (!(root -> left || root -> right)) return root -> val != 0;
        
        // Store evaluations for the left and right subtree.
        bool evaluateRoot;
        bool evaluateLeftSubtree = evaluateTree(root -> left), evaluateRightSubtree = evaluateTree(root -> right);
        if (root -> val == 2)
            evaluateRoot = evaluateLeftSubtree | evaluateRightSubtree;
        else
            evaluateRoot = evaluateLeftSubtree & evaluateRightSubtree;
        return evaluateRoot;
    }
};

File: 2331-evaluate-boolean-binary-tree/2331-evaluate-boolean-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def evaluateTree(self, root: Optional[TreeNode]) -> bool:
        stack, evaluated = [root], dict()
        
        while stack:
            top_node = stack[-1]
            
            # If the node is a leaf node, store its value in the evaluated dictionary
            # and continue
            if not (top_node.left or top_node.right):
                stack.pop()
                evaluated[top_node] = top_node.val == 1
                continue
                
            # If both the children have already been evaluated, use their
            # values to evaluate the current node.
            if top_node.left in evaluated and top_node.right in evaluated:
                stack.pop()
                if top_node.val == 2:
                    evaluated[top_node] = evaluated[top_node.left] or evaluated[top_node.right]
                else:
                    evaluated[top_node] = evaluated[top_node.left] and evaluated[top_node.right]
            else:
                # If both the children are not leaf nodes, push the current
                # node along with its left and right child back into the stack.
                if top_node.left:
                    stack.append(top_node.left)
                
                if top_node.right:
                    stack.append(top_node.right)
        
        return evaluated[root]
            